torchmdnet/neighbors/neighbors_cuda.cu

```diff
@@ -2,45 +2,79 @@
    Connection between the neighbor CUDA implementations and the torch extension.
    See neighbors.cpp for the definition of the torch extension functions.
  */
-#include <torch/extension.h>
 #include "neighbors_cuda_brute.cuh"
 #include "neighbors_cuda_cell.cuh"
 #include "neighbors_cuda_shared.cuh"
+#include <torch/extension.h>
+template <class... T> auto call_forward_kernel(const std::string& kernel_name, const T&... args) {
+    if (kernel_name == "brute") {
+        return forward_brute(args...);
+    } else if (kernel_name == "cell") {
+        return forward_cell(args...);
+    } else if (kernel_name == "shared") {
+        return forward_shared(args...);
+    } else {
+        throw std::runtime_error("Unknown kernel name");
+    }
+}
+
+// This is the autograd function that is called when the user calls get_neighbor_pairs.
+// It dispatches the required strategy for the forward function and implements the backward
+// function. The backward function is written in full pytorch so that it can be differentiated a
+// second time automatically via Autograd.
+class NeighborAutograd : public torch::autograd::Function<NeighborAutograd> {
+public:
+    static tensor_list forward(AutogradContext* ctx, const std::string& strategy,
+                               const Tensor& positions, const Tensor& batch,
+                               const Tensor& box_vectors, bool use_periodic,
+                               const Scalar& cutoff_lower, const Scalar& cutoff_upper,
+                               const Scalar& max_num_pairs, bool loop, bool include_transpose) {
+        Tensor neighbors, deltas, distances, i_curr_pair;
+        std::tie(neighbors, deltas, distances, i_curr_pair) =
+            call_forward_kernel(strategy, positions, batch, box_vectors, use_periodic, cutoff_lower,
+                                cutoff_upper, max_num_pairs, loop, include_transpose);
+        ctx->save_for_backward({neighbors, deltas, distances});
+        ctx->saved_data["num_atoms"] = positions.size(0);
+        return {neighbors, deltas, distances, i_curr_pair};
+    }
+
+    static tensor_list backward(AutogradContext* ctx, tensor_list grad_outputs) {
```

Comment:
Is this compatible with CUDA Graphs?

Reply:
It is not, sadly...
libtorch is lacking some convenient functions to facilitate that, like masked_index_add.
I am going to merge this and work on that in another PR.